browser-extension/mv3/src/content-scripts/client/ruleExecution.ts

```diff
@@ -12,23 +19,211 @@ export const initRuleExecutionHandler = () => {
+  testRuleWidget.setAttribute("applied-status", appliedRuleIds.has(ruleId).toString());
+  setWidgetInfoText(testRuleWidget, ruleDetails);
+
+  document.documentElement.appendChild(testRuleWidget);
+
+  testRuleWidget.addEventListener("view-results", () => {
+    chrome.runtime.sendMessage({
+      action: EXTENSION_MESSAGES.SAVE_TEST_RULE_RESULT,
+      ruleId,
+      appliedStatus: testRuleWidget?.getAttribute("applied-status") === "true",
+    });
+  });
+};
+
+const setWidgetInfoText = (testRuleWidget: HTMLElement, ruleDetails: Rule) => {
+  const { ruleType } = ruleDetails;
+
+  switch (ruleType) {
+    case "Response":
+      testRuleWidget.setAttribute(
+        "rq-test-rule-text",
+        `Response Modifications will not show up in the browser network devtools due to technical contraints. Checkout docs for more <a target="_blank" href="https://developers.requestly.io/http-rules/modify-response-body/">details</a>.`
+      );
+      break;
+    case "Headers":
+      const responseHeaderExists = ruleDetails.pairs.some((pair) => {
+        return pair?.modifications?.Response?.length > 0;
+      });
+      responseHeaderExists &&
+        testRuleWidget.setAttribute(
+          "rq-test-rule-text",
+          `Response Header Modifications will not show up in the browser network devtools due to technical constraints. Checkout docs for more <a target="_blank" href="https://developers.requestly.io/http-rules/modify-headers/">details</a>.`
+        );
+      break;
+    default:
+      return;
+  }
+};
+
+const notifyRuleAppliedToExplicitWidget = (ruleId: string) => {
+  const explicitTestRuleWidget = document.querySelector("rq-explicit-test-rule-widget");
+
+  if (explicitTestRuleWidget?.getAttribute("applied-status") === "false") {
+    if (explicitTestRuleWidget.getAttribute("rule-id") === ruleId) {
+      explicitTestRuleWidget.setAttribute("applied-status", "true");
+    }
+
+    explicitTestRuleWidget.dispatchEvent(
+      new CustomEvent("new-rule-applied", {
+        detail: {
+          appliedRuleId: ruleId,
+        },
+      })
+    );
+  }
+};
+
+export const showImplicitTestRuleWidget = async () => {
+  if (document.querySelector("rq-implicit-test-rule-widget")) {
+    return;
+  }
+
+  const testRuleWidget = document.createElement("rq-implicit-test-rule-widget");
+
+  testRuleWidget.classList.add("rq-element");
+  testRuleWidget.style.display = "none";
+
+  let appliedRules = [];
+  for (let ruleId of appliedRuleIds.values()) {
+    const ruleDetails = await getRule(ruleId);
+    if (shouldShowImplicitWidget(ruleDetails.ruleType)) {
+      appliedRules.push({
+        ruleId: ruleDetails.id,
+        ruleName: ruleDetails.name,
+        ruleType: ruleDetails.ruleType,
+      });
+    }
+  }
+  testRuleWidget.setAttribute("applied-rules", JSON.stringify(appliedRules));
+  document.documentElement.appendChild(testRuleWidget);
+
+  if (appliedRules.length) {
+    testRuleWidget.style.display = "block";
+  }
```

Comment:
Can we add a listener which listens for widget's `rule_applied_listener_active` and then call `handleAppliedRuleNotification` for enabled rules types from cache, instead  using `applied-rules` attributes?

The same way we have implemented in mv2 widget.